wandb/sdk/launch/sweeps/scheduler_optuna.py

```diff
@@ -0,0 +1,616 @@
+        for param, extras in self._sweep_config["parameters"].items():
+            if extras.get("values"):
+                config[param]["value"] = trial.suggest_categorical(
+                    param, extras["values"]
+                )
+            elif extras.get("value"):
+                config[param]["value"] = trial.suggest_categorical(
+                    param, [extras["value"]]
+                )
+            elif type(extras.get("min")) == float:
+                if not extras.get("max"):
+                    raise SchedulerError(
+                        f"{LOG_PREFIX}Error converting config. 'min' requires 'max'"
+                    )
+                log = "log" in param
+                config[param]["value"] = trial.suggest_float(
+                    param, extras["min"], extras["max"], log=log
+                )
+            elif type(extras.get("min")) == int:
+                if not extras.get("max"):
+                    raise SchedulerError(
+                        f"{LOG_PREFIX}Error converting config. 'min' requires 'max'"
+                    )
+                log = "log" in param
+                config[param]["value"] = trial.suggest_int(
+                    param, extras["min"], extras["max"], log=log
+                )
+            else:
+                logger.debug(f"Unknown parameter type: {param=}, {extras=}")
+        return config, trial
+
+    def _make_trial_from_objective(self) -> Tuple[Dict[str, Any], optuna.Trial]:
+        """
+        This is the core logic that turns a user-provided MOCK objective func
+            into wandb params, allowing for pythonic search spaces.
+            MOCK: does not actually train, only configures params
+
+        First creates a copy of our real study, quarantined from fake metrics
+
+        Then calls optuna optimize on the copy study, passing in the
+        loaded-from-user objective function with an aggresive timeout:
+            ensures the model does not actually train.
+
+        Retrieves created mock-trial from study copy and formats params for wandb
+
+        Finally, ask our real study for a trial with fixed params = retrieved
+
+        Returns wandb formatted config and optuna trial from real study
+        """
+        wandb.termlog(
+            f"{LOG_PREFIX}Making trial params from objective func, ignoring sweep config parameters"
+        )
+        study_copy = optuna.create_study()
+        study_copy.add_trials(self.study.trials)
+
+        # Signal handler to raise error if objective func takes too long
+        import signal
+
+        def handler(signum, frame):
+            raise TimeoutError(
+                "Passed optuna objective function only creates parameter config."
+                f" Do not train; must execute in {2} seconds. See docs."
+            )
+
+        signal.signal(signal.SIGALRM, handler)
+        signal.alarm(2)
+
+        study_copy.optimize(self._objective_func, n_trials=1)
```

Comment:
pretty sure a well-crafted malicious user-provided objective function might be able to affect this signal. Only a potential vulnerability if wandb starts into the scheduler-running business, which is unlikely.

Reply:
Also i'm so sorry for having to do this, apparently the `timeout` function that optuna provides only terminates after completed trials, and doesn't interrupt trials when over the time limit. Which is of course useless in our use-case.